unified-scheduler-logic/src/lib.rs

```diff
@@ -1,20 +1,1299 @@
+    /// and not creating more than one, onto the API consumers. And its constructor is non-`const`,
+    /// and the type is `!Clone` (and `!Copy` as well), `!Default`, `!Send` and `!Sync` to make it
+    /// relatively hard to cross thread boundaries accidentally.
+    ///
+    /// In other words, the token semantically _owns_ all of its associated instances of
+    /// [`TokenCell`]s. And `&mut Token` is needed to access one of them as if the one is of
+    /// [`Token`]'s `*_mut()` getters. Thus, the Rust aliasing rule for `UnsafeCell` can
+    /// transitively be proven to be satisfied simply based on the usual borrow checking of the
+    /// `&mut` reference of [`Token`] itself via [`::borrow_mut()`](TokenCell::borrow_mut).
+    ///
+    /// By extension, it's allowed to create _multiple_ tokens in a _single_ process as long as no
+    /// instance of [`TokenCell`] is shared by multiple instances of [`Token`].
+    ///
+    /// Note that this is overly restrictive in that it's forbidden, yet, technically possible
+    /// to _have multiple mutable references to the inner values at the same time, if and only
+    /// if the respective cells aren't aliased to each other (i.e. different instances)_. This
+    /// artificial restriction is acceptable for its intended use by the unified scheduler's code
+    /// because its algorithm only needs to access each instance of [`TokenCell`]-ed data once at a
+    /// time. Finally, this restriction is traded off for restoration of Rust aliasing rule at zero
+    /// runtime cost.  Without this token mechanism, there's no way to realize this.
+    #[derive(Debug, Default)]
+    pub(super) struct TokenCell<V>(UnsafeCell<V>);
+
+    impl<V> TokenCell<V> {
+        /// Creates a new `TokenCell` with the `value` typed as `V`.
+        ///
+        /// Note that this isn't parametric over the its accompanied `Token`'s lifetime to avoid
+        /// complex handling of non-`'static` heaped data in general. Instead, it's manually
+        /// required to ensure this instance is accessed only via its associated Token for the
+        /// entire lifetime.
+        ///
+        /// This is intentionally left to be non-`const` to forbid unprotected sharing via static
+        /// variables among threads.
+        pub(super) fn new(value: V) -> Self {
+            Self(UnsafeCell::new(value))
+        }
+
+        /// Returns a mutable reference with its lifetime bound to the mutable reference of the
+        /// given token.
+        ///
+        /// In this way, any additional reborrow can never happen at the same time across all
+        /// instances of [`TokenCell<V>`] conceptually owned by the instance of [`Token<V>`] (a
+        /// particular thread), unless previous borrow is released. After the release, the used
+        /// singleton token should be free to be reused for reborrows.
+        pub(super) fn borrow_mut<'t>(&self, _token: &'t mut Token<V>) -> &'t mut V {
```

Comment:
I understand the intent, but this is akin to a transmute and can be made
segfault trivially

```
struct T {
	v: Vec<u8>,
};

let mut token = unsafe { Token::<T>::assume_exclusive_mutating_thread() };
let c = {
	let cell = TokenCell::new(T { v: vec![42] });
	cell.borrow_mut(&mut token)
};
c.v.push(43);
```

Reply:
@alessandrod oh, quite nice catch.. hope this fixes segfault and no more ub?: 001b10e17d6444ba296c27d25191408f614bdf85

seems i'm too short-sighted only with aliasing-based ub.. ;)

Reply:
uggg, actually, i need `with_borrow_mut()`.. I'll soon push this.

Reply:
> uggg, actually, i need `with_borrow_mut()`.. I'll soon push this.

done: 02567b18b28e97bcb0d3319b190467f58b2b461d